express-notes-api/index.js

```diff
@@ -0,0 +1,115 @@
+const express = require('express');
+const app = express();
+const fs = require('fs');
+
+const content = require('./data.json');
+const notes = content.notes;
+let nextId = 10;
+
+app.use(express.json());
+
+app.get('/api/notes', (req, res) => {
+  const apiArray = [];
+  for (note in notes){
+    apiArray.push(notes[note]);
+  }
+  res.json(apiArray);
+});
+
+app.get('/api/notes/:id', (req, res) => {
+  const idProp = parseInt(req.params.id);
+
+  if (idProp < 0 || !Number.isInteger(idProp)) {
+    res.status(400).json({ 'error': 'id must be a positive integer' });
+  } else if (!notes[idProp]){
+    res.status(404).json({ 'error': `cannot find note with id ${idProp}` });
+  } else {
+    res.json(notes[idProp]);
+  }
+});
+
+app.post('/api/notes', (req, res) => {
+  const newNote = req.body;
+  let uid = nextId;
+
+  if (newNote.content){
+    notes[uid] = newNote;
+    newNote.id = nextId++;
+    res.status(201).json(newNote);
+
+    let data = JSON.stringify(content, null, 2);
+
+    fs.writeFile('data.json', data, (err) => {
+      if (err) {
+        throw err
+      };
+    });
```

Comment:
You don't want to send a response until everything has completed, including writing the data to the file. Should look something like:

```JAVASCRIPT
let data = JSON.stringify(content, null, 2);

fs.writeFile('data.json', data, (err) => {
  if (err) {
        return res.status(500).json({error: 'Unexpected server error'});
  };
  res.status(201).json(newNote);
});
```